src/core/execute.c

```diff
@@ -210,6 +210,66 @@ bool exec_needs_ipc_namespace(const ExecContext *context) {
         return context->private_ipc || context->ipc_namespace_path;
 }
 
+static bool can_apply_cgroup_namespace(const ExecContext *context, const ExecParameters *params) {
+        int r;
+
+        r = cg_all_unified();
+        if (r < 0) {
+                log_once_exec_warning_errno(context, params, r, "Failed to determine cgroup hierarchy version: %m");
+                return false;
+        }
+        if (r == 0) {
+                log_once_exec_warning(context, params,
+                                      "ProtectControlGroups=%s is configured, but the unified cgroups hierarchy is not set up, ignoring namespace setup.",
+                                      protect_control_groups_to_string(context->protect_control_groups));
+                return false;
+        }
+
+        if (!ns_type_supported(NAMESPACE_CGROUP)) {
+                log_once_exec_warning(context, params,
+                                      "ProtectControlGroups=%s is configured, but the kernel does not support cgroup namespaces, ignoring namespace setup.",
+                                      protect_control_groups_to_string(context->protect_control_groups));
+                return false;
+        }
+        return true;
+}
+
+static bool needs_cgroup_namespace(ProtectControlGroups i) {
+        return IN_SET(i, PROTECT_CONTROL_GROUPS_PRIVATE, PROTECT_CONTROL_GROUPS_STRICT);
+}
+
+ProtectControlGroups exec_get_protect_control_groups(const ExecContext *context, const ExecParameters *params) {
+        /* If cgroup namespace is configured via ProtectControlGroups=private or strict but we can't actually
+         * use cgroup namespace, either from not having unified hierarchy or kernel support, we ignore the
+         * setting and do not unshare the namespace. ProtectControlGroups=private and strict get downgraded
+         * to no and yes respectively. This ensures that strict always gets a read-only mount of /sys/fs/cgroup. */
+        if (needs_cgroup_namespace(context->protect_control_groups) && !can_apply_cgroup_namespace(context, params)) {
```

Comment:
I.e. callers of a getter could potentially trigger a warning log, which seems absurd. Let's just document this and fall back gracefully. Might as well leave a TODO for removal in v258.

Reply:
Fallback is already documented in xml files and comments. Will leave a TODO